File: apps/Skin-Image-Classification/iOS/SkinImageClassification/Models/Classification.swift
import Foundation

/// One ranked (class, probability) pair from the classifier's softmax output.
struct ClassScore: Identifiable {
    let skinClass: SkinClass
    let probability: Float
    var id: Int { skinClass.rawValue }
}

/// The classifier's verdict for a single photo: the full softmax distribution
/// (ranked, highest first) plus inference latency.
struct Classification {
    /// All 7 classes ranked by probability, highest first.
    let ranked: [ClassScore]
    /// On-device inference time in milliseconds.
    let latencyMs: Double

    var top: ClassScore { ranked[0] }
    var topClass: SkinClass { top.skinClass }
    var confidence: Float { top.probability }

    /// Runner-up classes (everything below the top), for the distribution bars.
    var runnersUp: [ClassScore] { Array(ranked.dropFirst()) }

    /// Below this the model is effectively guessing — UI and MedGemma both soften.
    var isLowConfidence: Bool { confidence < 0.60 }

    /// Build from raw logits in `SkinClass.allCases` order.
    init(logits: [Float], latencyMs: Double) {
        let probs = Self.softmax(logits)
        ranked = SkinClass.allCases
            .map { ClassScore(skinClass: $0, probability: probs[safe: $0.rawValue] ?? 0) }
            .sorted { $0.probability > $1.probability }
        self.latencyMs = latencyMs
    }

    private static func softmax(_ x: [Float]) -> [Float] {
        guard let m = x.max() else { return x }
        let e = x.map { Foundation.exp($0 - m) }
        let sum = e.reduce(0, +)
        return sum > 0 ? e.map { $0 / sum } : e
    }
}

private extension Array {
    subscript(safe index: Int) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}

File: apps/Skin-Image-Classification/Android/app/src/main/java/ai/zetic/skinclassifier/model/Classification.kt
package ai.zetic.skinclassifier.model

import ai.zetic.skinclassifier.core.AppConfig
import ai.zetic.skinclassifier.core.MelangeKit

/** One class with its softmax probability. */
data class ClassScore(val skinClass: SkinClass, val probability: Float)

/**
 * Result of running the classifier on one image: every class ranked by probability plus the
 * on-device latency. 1:1 with the iOS `Classification` struct.
 */
data class Classification(
    val ranked: List<ClassScore>,
    val latencyMs: Double,
) {
    val top: ClassScore get() = ranked.first()
    val topClass: SkinClass get() = top.skinClass
    val confidence: Float get() = top.probability
    val isLowConfidence: Boolean get() = confidence < AppConfig.LOW_CONFIDENCE

    companion object {
        /** Build from raw logits (length >= 7): softmax, then rank classes descending. */
        fun fromLogits(logits: FloatArray, latencyMs: Double): Classification {
            val probs = MelangeKit.softmax(logits.copyOf(SkinClass.ordered.size))
            val ranked = SkinClass.ordered
                .mapIndexed { i, c -> ClassScore(c, probs[i]) }
                .sortedByDescending { it.probability }
            return Classification(ranked, latencyMs)
        }
    }
}
